---
title: "HPA Normalalized Tissue Expression Notebook"
output: html_notebook
---

```{r setup}
#install.packages("neurocInstall")
#library(neurocInstall)
#neuro_install('gganatogram')
library(gganatogram)
library(tidyverse)
library(here)
library(janitor)
library(cowplot)
#rm(list = ls())
```

#import RNA consensus tissue gene data
Consensus transcript expression levels summarized per gene in 74 tissues based on transcriptomics data from three sources: HPA, GTEx and FANTOM5. The consensus normalized expression ("NX") value is calculated as the maximum NX value for each gene in the three data sources. For tissues with multiple sub-tissues (brain regions, blood cells, lymphoid tissues and intestine) the maximum of all sub-tissues is used for the tissue type. The tab-separated file includes Ensembl gene identifier ("Gene"), analysed sample ("Tissue") and normalized expression ("NX"). The data is based on The Human Protein Atlas version 19.2 and Ensembl version 92.38.

You can either download the file and place it in the data folder
1. Download https://www.proteinatlas.org/download/rna_tissue_consensus.tsv.zip to the data folder  
2. Unzip (double-click)  
3. Confirm you see rna_tissue_consensus.tsv  

Or you can run the code below to download the file each time

```{r}

url <- "https://www.proteinatlas.org/download/rna_tissue_consensus.tsv.zip"
zip_file <- tempfile(fileext = ".zip")
download.file(url, zip_file, mode = "wb")
normalized_tissue <- read_tsv(zip_file) %>% 
  clean_names()

#uncomment this line to import from saved location
#normalized_tissue <- read_tsv(here::here("data", "rna_tissue_consensus.tsv")) %>% clean_names()

head(normalized_tissue)

count(normalized_tissue, nx, sort = TRUE)

ggplot(normalized_tissue) +
  geom_histogram(aes(x = log10(nx)))
```

#yfg
```{r}
normalized_tissue %>% 
  filter(gene_name == "NOS2") %>% 
  arrange(desc(nx))

normalized_tissue %>% 
  filter(tissue == "amygdala") %>% 
  top_n(20) %>% 
  arrange(desc(nx)) %>% 
  ggplot() +
  geom_col(aes(x = fct_reorder(gene_name, nx), y = nx)) +
  coord_flip() + 
  labs(title = "Tissue distribution of amygdala expression", x = "Normalized Expression", y = "") +
  theme_cowplot()


```


#gganatogram

```{r}
#On gganatograms site, one of the first things they instruct is as follows:
#"In order to use the function gganatogram, you need to have a data frame with organ, colour, and value"

organPlot <- data.frame(organ = c("heart", "leukocyte", "nerve", "brain", "liver", "stomach", "colon"), 
 type = c("circulation", "circulation",  "nervous system", "nervous system", "digestion", "digestion", "digestion"), 
 colour = c("red", "red", "blue", "blue", "purple", "orange", "orange"), 
 value = c(10, 5, 1, 8, 2, 5, 5), 
 stringsAsFactors=F)

#Let's take a look!
head(organPlot)
```

Alternatively, we can import a dataset already within gganatogram
```{r}
#First, we'll look at what datasets are available within the package
library(help="gganatogram")

#We're going to plot to a human male (since our dataset is human)
data(hgMale_key)

glimpse(hgMale_key)

```

Now we need to prepare our data to be compatable with gganatogram
```{r}
normalized_tissue$organ <- 0 #this is creating a new column
normalized_tissue$organ <- str_replace(normalized_tissue$tissue, " ", "_") #reformatting to fit hgMale_key format

#now we'll join the hgMale_key with our dataset by our new variable
norm_tiss_plot <- full_join(normalized_tissue, hgMale_key, by= c("organ"))

nx_plot<- norm_tiss_plot %>% select(- value) #removing value from original hgMale_key data

nx_plot$value <- nx_plot$nx #here I copied our normal expression values to the new "value" column, because this is the information gganatogram needs. I could have just renamed the column, but here it is just copied. 
```

```{r}

#filter our data for your gene of interest
T5_plot<- nx_plot %>% 
  filter(gene_name == "SIRT5") %>% 
  filter(!is.na(type)) %>% #sorting out any that do not have tissue type association
  arrange(desc(nx)) 

#First, let's see what gganatogram does by default
gganatogram(data=T5_plot, organism='human', sex='male', fill="value")
```

Let's start to customize our plot. To see gganatogram information, you can go to this site: https://jespermaag.github.io/blog/2018/gganatogram/ 

```{r}
#first, let's change the color of the human base
gganatogram(data=T5_plot, fillOutline='white', organism='human', sex='male', fill="value")

#now we're going to remove the outline, including face
gganatogram(data=T5_plot, fillOutline='white', outline= F, organism='human', sex='male', fill="value")
```

We can add common ggplot functions to our anatogram
```{r}
#Let's begin with the theme
gganatogram(data=T5_plot, fillOutline='white', outline= F, organism='human', sex='male', fill="value") + 
  theme_void()

#And color
gganatogram(data=T5_plot, fillOutline='white', outline= F, organism='human', sex='male', fill="value") + 
  theme_void()+
  scale_fill_gradient(low = "white", high = "red")

#Note, you can even use the viridis package for this scaling. Just use the scale_color_viridis() function. (And don't forget to load the viridis library)


#To finish off our graph, let's go ahead and add some labels                      
gganatogram(data=T5_plot, fillOutline='white', outline= F, organism='human', sex='male', fill="value") + 
  theme_void()+
  scale_fill_gradient(low = "white", high = "red")+
  labs(title= "Sirt5 tissue expression", caption = "Sirt5 expression is highest in skeletal muscle and liver.")

#Now you have a great gganatogram, let's save it!
ggsave(here::here("2020_03", "results", "SIRT5anatogram.png"), last_plot())


#Fix ratios
ggsave(here::here("2020_03", "results", "SIRT5anatogram.png"), last_plot(), units = "in", width = 5, height = 7)

```

If you enjoyed the gganatograms, know that they also make plots for mice and other organisms!




#Plotting heatmaps

```{r heatmap}
if (!requireNamespace("BiocManager", quietly = TRUE))
    install.packages("BiocManager")
BiocManager::install("ComplexHeatmap") # Run these three lines just once

library(ComplexHeatmap)

heatmap_df <- normalized_tissue %>% 
                          pivot_wider(.,
                          values_from = nx,
                          names_from = tissue) 

heatmap_df[is.na(heatmap_df)] <- 0 # for illustration purposes only! Imputation can be a fraught process
heatmap_matrix <- as.matrix(heatmap_df[ , c(-1, -2)])
heatmap_matrix <- scale(heatmap_matrix) # if you don't scale, the pancreas will form a cluster of its own!
# This is because the pancreas shows such high expression compared to other organs

# Heatmap of gene expression across vaarious tissues
# No tissue clusters
Heatmap(heatmap_matrix, cluster_rows = FALSE, show_row_names = FALSE, show_heatmap_legend = FALSE,
        border = TRUE, row_title = "Gene", column_title = "Tissue", column_title_side = "bottom",
        cluster_columns = TRUE)

# set the seed of R's random number generator
set.seed(42) # reproducible randomness. K-means is a stochastic algorithm
# Two clusters
Heatmap(heatmap_matrix, cluster_rows = FALSE, show_row_names = FALSE, show_heatmap_legend = FALSE,
        border = TRUE, row_title = "Gene", column_title = "Tissue", column_title_side = "bottom",
        cluster_columns = TRUE, column_km = 2) # Brain tissues cluster together!

# Three clusters
Heatmap(heatmap_matrix, cluster_rows = FALSE, show_row_names = FALSE, show_heatmap_legend = FALSE,
        border = TRUE, row_title = "Gene", column_title = "Tissue", column_title_side = "bottom",
        cluster_columns = TRUE, column_km = 3)

# And so on

# As a challenge, repeat this with genes
```











